rust/relay/ebpf-turn-router/src/main.rs

```diff
@@ -164,270 +164,1385 @@ fn try_handle_turn_ipv4(ctx: &XdpContext, eth: Eth) -> Result<(), Error> {
+
+    // SAFETY: This is the only mutable instance of `UdpHdr` in this scope.
+    let udp = unsafe { ref_mut_at::<UdpHdr>(ctx, EthHdr::LEN + Ipv6Hdr::LEN)? };
+    udp.set_source(new_udp_src);
+    udp.set_dest(new_udp_dst);
+    udp.set_len(new_udp_len);
+
+    // Incrementally update UDP checksum
+
+    // TODO: Remove conditional checksums once we can test this fully in CI
+    if !crate::config::udp_checksum_enabled() {
+        udp.set_check(0);
+    } else {
+        let check = ChecksumUpdate::new(old_udp_check)
+            .remove_u32(u32::from_be_bytes(old_ipv4_src.octets()))
+            .remove_u32(u32::from_be_bytes(old_ipv4_dst.octets()))
+            .remove_u16(old_udp_src)
+            .remove_u16(old_udp_dst)
+            .remove_u16(old_udp_len)
+            .remove_u16(old_udp_len)
+            .remove_u16(channel_number)
+            .remove_u16(channel_data_length)
+            .add_u128(u128::from_be_bytes(new_ipv6_src.octets()))
+            .add_u128(u128::from_be_bytes(new_ipv6_dst.octets()))
+            .add_u16(new_udp_src)
+            .add_u16(new_udp_dst)
+            .add_u16(new_udp_len)
+            .add_u16(new_udp_len)
+            .into_checksum();
+
+        if check == 0 {
+            udp.set_check(0xFFFF); // Special case for zero checksum - write 0xFFFF to the wire
+        } else {
+            udp.set_check(check);
         }
+    }
 
-        Error::NoEntry(SupportedChannel::ChanToUdp66)
-    })?;
+    Ok(())
+}
 
-    let new_src = ipv6.dst(); // The IP we received the packet on will be the new source IP.
-    let new_dst = port_and_peer.peer_ip();
-    let new_ipv6_payload_len = ipv6.payload_len() - CdHdr::LEN as u16;
+#[inline(always)]
+fn handle_ipv6_udp_to_ipv6_channel(
+    ctx: &XdpContext,
+    client_and_channel: &ClientAndChannelV6,
+) -> Result<(), Error> {
+    // Expand by 4 bytes for channel data header
+    const NET_EXPANSION: i32 = -(CdHdr::LEN as i32);
 
-    eth.swap_macs()?;
+    adjust_head(ctx, NET_EXPANSION)?;
 
-    let pseudo_header = ipv6.update(new_src, new_dst, new_ipv6_payload_len);
+    // Now read the old packet data from its NEW location (shifted by 4 bytes)
+    let old_data_offset = CdHdr::LEN;
 
-    let new_udp_len = udp.len() - CdHdr::LEN as u16;
-    udp.update(
-        pseudo_header,
-        port_and_peer.allocation_port(),
-        port_and_peer.peer_port(),
-        new_udp_len,
-        cd.number(),
-        cd.length(),
-    );
+    let (old_eth_src, old_eth_dst, old_eth_type) = {
+        // SAFETY: This is the only mutable instance of `EthHdr` in this scope.
+        let old_eth = unsafe { ref_mut_at::<EthHdr>(ctx, old_data_offset)? };
+        (old_eth.src_addr, old_eth.dst_addr, old_eth.ether_type)
+    };
+
+    let (
+        old_ipv6_src,
+        old_ipv6_dst,
+        old_ipv6_len,
+        old_ipv6_priority,
+        old_ipv6_flow_label,
+        old_ipv6_hop_limit,
+        old_ipv6_next_hdr,
+    ) = {
```

Comment:
We know it must be UDP, why not hardcode it?

Reply:
I chose to stick to copying the fields explicitly from the old packet to maximize the chance the compiler is able to spot a contiguous region of memory and do a wider memmove.

I have doubts it'll perform that optimization but I'd like to stick to this approach so that if / when we come back and do another round of optimization, it'll be easier to spot which fields we can move directly and which ones we need to pull out and operate on.